src/core.tsx

```diff
@@ -169,9 +169,20 @@ export function createShikiTextarea(styles: Record<string, string>) {
               spellcheck={false}
               class={styles.textarea}
               disabled={!config.editable}
-              onInput={e => {
-                const value = e.currentTarget.value
+              on:input={e => {
+                const target = e.currentTarget
+                const value = target.value
+
+                // local
                 setSource(value)
+
+                // copy to custom element document fragment
+                const rootNode = target.getRootNode()
+                if (rootNode && rootNode instanceof ShadowRoot) {
+                  rootNode.value = value
+                }
+
+                // user provided callback
```

Comment:
This looks like state management is needed. Instead of this, because it is difficult o trace and figure out where the value is coming from in the code that needs the value, I would set a signal (or a store property).

That signal or store property could be from a context, for example, or passed downward from the higher component (f.e. the custom element). This way the code will be clear and also type checked.

Reply:
Oh, I see you already have `config.onInput` for example. You could use that, or similar, from the custom element.